# Genome assembly using Nanopore reads

## Description
Nanopore reads generated from Oxford Nanopore Technologies (ONT) are much longer than the reads produced by second-generation sequencing technologies, therefore can improve the accuracy of genome assemblies by spanning the highly-repetitive segments that cause difficulty. 
Today's long-read technologies are more accurate and less expensive, making them the choice for the assembly of complex genomes. 
Nanopore sequencing has been largely applied to assemble high-quality genomes in GA. 
This space is for documenting experiences of using programs designed for genome assembly using nanopore reads, including ideas on parameter setting and tips for running programs. 
All are welcome to contribute. 

## Bioinformaticians
**Dr. Chen Wu**<BR>
Plant and Food Research<BR>
Species: Bilberry and Gillenia

...

## Literature & resource
* Assembler
    * [A list of 88 free genome assemblers](https://bioinformaticshome.com/tools/wga/wga.html)
* Benchmarking ideas
    * [prokaryote whole genome sequencing](https://github.com/rrwick/Long-read-assembler-comparison)
* Review
    * [Oxford Nanopore Sequencing: New opportunities for plant genomics?](https://academic.oup.com/jxb/advance-article/doi/10.1093/jxb/eraa263/5847836)
