# Genome assembly from nanopore reads and plus

A space	for documenting	ideas, methods plus tips on nanopore read assembly

## Description
Nanopore reads generated from Oxford Nanopore Technologies (ONT) are much longer than the reads produced by second-generation short-read technologies, therefore can improve the accuracy of genome assemblies by spanning the highly-repetitive segments that cause difficulty. Today's long-read technologies are more accurate and less expensive, making them the choice for the assembly of complex genomes. Nanopore sequencing has been largely applied to assemble high-quality genomes in GA, hense a review on methods/approaches for long read QC, assembler benchmarking as well as assembly quality evaluation is required. We also include ideas on parameter setting and tips for running programs in this folder for discussion. All are welcome to contribute. 

## Bioinformaticians
**Dr. Chen Wu**
Plant and Food Research
Species: Bilberry and Gillenia

...

## Literature & resource
* Assembler
    * [A list of 88 free genome assemblers](https://bioinformaticshome.com/tools/wga/wga.html)
* Benchmarking ideas
    * [prokaryote whole genome sequencing](https://github.com/rrwick/Long-read-assembler-comparison)
* Review
    * [Oxford Nanopore Sequencing: New opportunities for plant genomics?](https://academic.oup.com/jxb/advance-article/doi/10.1093/jxb/eraa263/5847836)
